frame/system/src/migrations/mod.rs

```diff
@@ -0,0 +1,182 @@
+type RefCount = u32;
+
+/// All balance information for an account.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Default, RuntimeDebug, MaxEncodedLen)]
+struct AccountData<Balance> {
+	free: Balance,
+	reserved: Balance,
+	misc_frozen: Balance,
+	fee_frozen: Balance,
+}
+
+/// Information of an account.
+#[derive(Clone, Eq, PartialEq, Default, RuntimeDebug, Encode, Decode)]
+struct AccountInfo<Index, AccountData> {
+	nonce: Index,
+	consumers: RefCount,
+	providers: RefCount,
+	sufficients: RefCount,
+	data: AccountData,
+}
+
+/// Trait to implement to give information about types used for migration
+pub trait V2ToV3 {
+	/// The elections-phragmen pallet.
+	type Pallet: 'static + PalletInfoAccess;
+
+	/// System config account id
+	type AccountId: 'static + FullCodec;
+
+	/// System config index
+	type Index: 'static + FullCodec + Copy;
+
+	/// Currency balance.
+	type Balance: 'static + FullCodec + Copy;
+}
+
+frame_support::generate_storage_alias!(
+	System, UpgradedToU32RefCount => Value<
+		bool,
+		ValueQuery
+	>
+);
+
+frame_support::generate_storage_alias!(
+	System, UpgradedToTripleRefCount => Value<
+		bool,
+		ValueQuery
+	>
+);
+
+frame_support::generate_storage_alias!(
+	System, Account<T: V2ToV3> => Map<
+		(Blake2_128Concat, T::AccountId),
+		AccountInfo<T::Index, AccountData<T::Balance>>
+	>
+);
+
+/// Apply the migrations from 2 to 3.
+///
+/// ### Warning
+///
+/// This function tries to infer the right migration to apply only based on whether it could
+/// decode any value consistent with a state representation.
+///
+/// + If the state is fully migrated, it does nothing.
+/// + If the state is in single u8 reference counting, it migrates the full state to triple
+/// (u32) reference counting
+/// + Otherwise, it first tries to migrate assuming the state is dual reference counting,
+/// and *only if no element has been translated* tries to apply the migration from single
+/// (u32) to triple reference counting.
+///
+/// If unsure about the behaviour, the standalone functions below have to be preferred.
+pub fn apply<T: V2ToV3>() -> Weight {
+	if !<UpgradedToTripleRefCount>::get() {
+		if !<UpgradedToU32RefCount>::get() {
+			let _ = migrate_from_single_u8_to_triple_ref_count::<T>();
+			<UpgradedToU32RefCount>::put(true);
+		} else {
+			let nb_translated = migrate_from_dual_to_triple_ref_count::<T>();
+			if 0 == nb_translated {
+				let _ = migrate_from_single_to_triple_ref_count::<T>();
+			}
+		}
+		<UpgradedToTripleRefCount>::put(true);
+		Weight::max_value()
+	} else {
+		log::info!(
+			target: "runtime::system",
+			"Already upgraded to triple reference counting. No migration necessary."
+		);
+		0
+	}
+}
+
+use super::*;
+
+#[allow(dead_code)]
+/// Migrate from unique `u8` reference counting to triple `u32` reference counting.
+pub fn migrate_from_single_u8_to_triple_ref_count<T: V2ToV3>() -> usize {
+	let mut translated: usize = 0;
+	<Account<T>>::translate::<(T::Index, u8, AccountData<T::Balance>), _>(
```

Comment:
this assume that account data is storing balance, this is not necessarily the case.

Reply:
Doesn't it? This is part of the struct definition, and it seems the previous migration logic also assumed this.

Reply:
no previous translate was:
```
Account::<T>::translate::<(T::Index, u8, T::AccountData), _>
```
So you see it uses the generic type `T::AccountData`, here you use `AccountData<T::Balance>` which one possible concrete type for the generic type, but there could be other concrete type, such as `()`.

Reply:
Ok I understand what you mean now. I looked at AccountData definition, and most of the (non unit) definitions are balances, but formally, it is meant to be anything. 
Then, what I did was to remove the definition of AccountData altogether, and remove the balance from the AccountInfo, instead, the AccountData will need to be provided in the runtime when implementing the binding trait V2ToV3.